test/azure_devops_agent_aci_test.go

```diff
@@ -284,12 +291,90 @@ func TestDeployAzureDevOpsLinuxAgentsIntoExistingRresourceGroup(t *testing.T) {
 			t.Fatalf("Test failed. Expected number of agents is %d. Actual number of agents is %d", expectedAgentsCount, actualAgentsCount)
 		}
 	})
+}
+
+// This function tests the deployment of Azure DevOps Linux and Windows agents
+func TestAgentCleanUp(t *testing.T) {
+	t.Parallel()
+
+	fixtureFolder := "./fixture/agent-pool-cleanup"
+
+	// generate a random suffix for the test
+	rand.Seed(time.Now().UnixNano())
+	randomInt := rand.Intn(9999)
+	randomSuffix := strconv.Itoa(randomInt)
+
+
+	// create random agent pool name
+	testPoolName := fmt.Sprintf("e2e-agents-%s", randomSuffix)
+	os.Setenv("TF_VAR_linux_azure_devops_pool_name", testPoolName)
+	os.Setenv("TF_VAR_windows_azure_devops_pool_name", testPoolName)
+
+	devopsOrganizationName := os.Getenv("TF_VAR_azure_devops_org_name")
+	devopsPersonalAccessToken := os.Getenv("TF_VAR_azure_devops_personal_access_token")
+	devopsOrganizationURL := fmt.Sprintf("https://dev.azure.com/%s", devopsOrganizationName)
+
+	// create the Linux agents pool
+	defer deleteAzureDevOpsAgentTestPool(testPoolName, devopsOrganizationURL, devopsPersonalAccessToken)
+	err := createAzureDevOpsAgentTestPool(testPoolName, devopsOrganizationURL, devopsPersonalAccessToken)
+	if err != nil {
+		t.Fatalf("Cannot create Azure DevOps Linux agent pool for the test: %v", err)
+	}
+
 
 	// At the end of the test, clean up any resources that were created
-	test_structure.RunTestStage(t, "teardown", func() {
+	defer test_structure.RunTestStage(t, "teardown", func() {
```

Comment:
Don't you think it makes sense to not call it with defer but add another test stage after the teardown that actually validates the agent pool is back to 0 agent, before destroying the pool?

Reply:
I added the defer because if the test fails the Stage will not be called. I did not want to have resources left in the Azure subscription.  

How should be handle that?

Reply:
Good question. Can we have a terraform destroy before the defered call is made? And check the pool in between? So it will ensure the pool is cleaned up, but still call the defered destroy (that will do nothing, because already cleaned up). But in case of something fails during the test, the defered destroy will still be useful. Makes sense?

Reply:
I can make whatever changes you want but I am new to this so I would need you to provide an example or to make the change

Reply:
I added a suggestion below, it needs to be tested, but the idea is to call the terraform destroy before the end of validate stage, check the number of agent declared in the pool is 0 after the destroy. Then the teardown (which is still defered) will run to cleanup anything that could be remaining.

Reply:
@mozts2005 have you had a chance to look at my suggestion? Anything blocking on your side? Let me know if I can help on something. Thanks :)